### PR TITLE
docs(layout): naive copy-and-paste of legacy layout docs content

### PR DESCRIPTION
**Pull request by apetro**
Sharing this WIP branch soonest to signal that I'm working on this bit of documentation to de-conflict.

##### Checklist

-   [ ] the [individual contributor license agreement][] is signed
-   [ ] commit message follows [commit guidelines][]
-   [ ] tests are included
-   [ ] documentation is changed or added
-   [ ] [message properties][] have been updated with new phrases
-   [ ] view conforms with [WCAG 2.0 AA][]

##### Description of change

[individual contributor license agreement]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#individual-contributor-license-agreement
[commit guidelines]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#commit
[message properties]: https://github.com/Jasig/uPortal/tree/master/uportal-war/src/main/resources/properties/i18n
[WCAG 2.0 AA]: https://www.w3.org/WAI/WCAG20/quickref/?levels=aaa&technologies=smil%2Cpdf%2Cflash%2Csl

### CODE REVIEW

File: docs/implement/layout-management/README.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    Delete Tab
+        If deleting the tab is restricted by a fragment owner then users will not be able to delete this tab from their layout. If a user has previously deleted this tab and the fragment owner then marks the tab to prevent deleting then the next time that the user logs in, that tab will reappear. When such a tab is restricted from being deleted the "Delete this tab" link is removed from the user interface for that tab when end users are editing their layout.
+
+Columns
+
+The action-restricting controls for columns are nearly identical to those of tabs as is their application.
+
+    Move Column
```
Comment (ChristianMurphy) [inline docs/implement/layout-management/README.md]: Is this intended to be a code block?

File: docs/implement/layout-management/README.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    Add Channels
+        If adding channels is restricted by a fragment owner then users will not be able to add channels to that column in their layout. If a user had previously added channels to that column and the fragment owner then marked the column to prevent adding channels then the next time that the user logs in those user added channels will be discarded. When such a column is restricted from adding channels, all "Add Channel" buttons are removed from the user's view of that column.
+    Delete Column
+        If deleting the column is restricted by a fragment owner then users will not be able to delete this column from their layout. Additionally, the user will not be able to delete the tab containing this column. When such a column is restricted from being deleted the "Delete this column" link is removed from the user interface for that column and the "Delete this tab" link is removed for its containing tab when end users are editing their layout.
+
+Channels
+
+The action-restricting controls for channels vary from those of columns and tabs in that only move and delete are restricted.
+
+    Move Channel
```
Comment (ChristianMurphy) [inline docs/implement/layout-management/README.md]: Is this intended to be a code block?

File: docs/implement/layout-management/README.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+The 'allowExpandedContent' Property
+
+For uPortal 3.2.5 and later, DLM was enhanced to support Expanded Content: for fragments: these are portlets outside of tabs – typically in the header and/or footer area – that get passed on to fragment audience members in the same way that tabs do. The allowExpandedContent feature is disabled by default in uPortal 3.2.x, but will be enabled (by default) in the next minor release of uPortal.
+
+DLM Expanded Content is a powerful and compelling approach to managing non-tab content for all the same reasons that DLM fragments are better than Template Users for managing administrator-provided tab content. Specifically, it's very difficult to make changes to Template User content once individual users have their own layouts in the database. For this reason, uPortal admins should provision content to users through DLM only; individual users can still add, move, and remove layout content for their own layouts as before.
+
+Warning
```
Comment (ChristianMurphy) [inline docs/implement/layout-management/README.md]: adding a :warning: `:warning:` can give the next some addition emphasis.

File: docs/implement/layout-management/make-a-fragment.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+Source: https://wiki.jasig.org/display/UPM43/Setup+a+fragment+layout+by+modifying+fragment-layout+XML+files
+
+This page is a work in progress.... It's in pretty good shape at this point though.
+
+You can create a new fragment layout by modify creating a new fragment-layout XML file, or alter an existing fragment layout by modifying an existing fragment-layout XML file.  When done, you import the fragment using one of the data import approaches (see Importing and Exporting data).
+
+The fragment-layout XML files are a hierarchy of folders and channels with some additional attributes and elements to help define the page structure.  An example is:
+Example fragment layout file
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ChristianMurphy) [inline docs/implement/layout-management/make-a-fragment.md]: :art: Code block it :slightly_smiling_face: 

    ```xml
    <?xml version="1.0" encoding="UTF-8"?>
    <layout>
      ...
    </layout>
    ```

File: docs/implement/layout-management/README.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+Although dlm.xml is xml based it does not have a defined schema or DTD. Its expected structure is defined by the expectations of the DLM loading infrastructure. This is in part due to the extensible nature of dlm.xml as will be seen in the declarations of the fragment audience definitions. Therefore, the remaining information in this section and related sections will portray the implicit structure expected by the DLM loading infrastructure.
+
+The dlm.xml file is located in the properties directory. The top level element in dlm.xml is the <managedLayoutFragments> element. It includes the namespace declaration for DLM used to prefix all expected DLM property, fragment, and audience elements. This element appears as follows in dlm.xml.<managedLayoutFragments xmlns:dlm="http://org.jasig.portal.layout.dlm.config">
```
Comment (ChristianMurphy) [inline docs/implement/layout-management/README.md]: :art: code blocks `<managedLayoutFragments>` :slightly_smiling_face:

File: docs/implement/layout-management/README.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+DLM Restrictions
+
+For this, an extended version of SimpleLayoutManagement's Preferences channel is provided and is the channel seen when using DLM and the Preferences link is selected. For non-fragment users they see that channel in the traditional manner. For fragment owner the preferences channel changes slightly to indicate the name of the fragment that is being edited as can be seen in the images below. Additionally, controls are added to the user interface to provide the ability to restrict what users of the fragment can do to fragment elements.
+Tabs
+
+In the image below the typical controls for a tab are shown along with some DLM specific fragment-user-actions that can be restricted. By default all check boxes are selected and hence all actions are allowed by end users. As seen in the image, the "Useful News" tab in the News fragment's layout restricts movement by end users of the fragment. To modify any of these actions the appropriate checkbox should be checked or unchecked as needed and the "Set Actions" button pressed. Each of these actions that can be restricted for tabs is discussed below.
+
+    Move Tab
```
Comment (ChristianMurphy) [inline docs/implement/layout-management/README.md]: Is this intended to be a code block?

File: docs/implement/layout-management/README.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+Warning
+
+Earlier distributions of uPortal, even distributions of uPortal 3.2, included the same header and footer content in layouts of every variety: users, template users, and fragment owners. This data is incompatible with the allowExpandedContent feature because it will cause users to have the same portlets several times.
+
+This result would have negative effects both visually and on performance. It was for this reason that allowExpandedContent was disabled by default in uPortal 3.2. If you want to use it, you must first make data adjustments to prevent these issues from occuring. The easiest way is to remove all existing header and footer portlets from DLM fragment owner layouts and from the fragmentTemplate layout. It is vital to make sure that users don't get the same header/footer portlet(s) from both template users and DLM fragments.
+
+Future Direction
```
Comment (ChristianMurphy) [inline docs/implement/layout-management/README.md]: Heading?

File: docs/implement/layout-management/README.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    Add Channels
+        If adding channels is restricted by a fragment owner then users will not be able to add channels to that column in their layout. If a user had previously added channels to that column and the fragment owner then marked the column to prevent adding channels then the next time that the user logs in those user added channels will be discarded. When such a column is restricted from adding channels, all "Add Channel" buttons are removed from the user's view of that column.
+    Delete Column
+        If deleting the column is restricted by a fragment owner then users will not be able to delete this column from their layout. Additionally, the user will not be able to delete the tab containing this column. When such a column is restricted from being deleted the "Delete this column" link is removed from the user interface for that column and the "Delete this tab" link is removed for its containing tab when end users are editing their layout.
+
+Channels
```
Comment (ChristianMurphy) [inline docs/implement/layout-management/README.md]: Heading?

File: docs/implement/layout-management/README.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+The dlm.xml file is located in the properties directory. The top level element in dlm.xml is the <managedLayoutFragments> element. It includes the namespace declaration for DLM used to prefix all expected DLM property, fragment, and audience elements. This element appears as follows in dlm.xml.<managedLayoutFragments xmlns:dlm="http://org.jasig.portal.layout.dlm.config">
+
+The only two supported child elements for managedLayoutFragments are <dlm:property> and <dlm:fragment>. Others may be included but they will be ignored. The <dlm:property> element has two required attributes, 'name' and 'value'. There are currently only two properties supported by DLM and if not defined, suitable defaults will be used as explained below.
+
+DLM Restrictions
```
Comment (ChristianMurphy) [inline docs/implement/layout-management/README.md]: Heading?

File: docs/implement/layout-management/make-a-fragment.md [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+Elements of type channel are publications of portlets on the page.  A channel may appear multiple times on the page, though it is generally not recommended.
+ID
+
+ID value. By convention folders are "s#" and channels are "n#" where # is a numeric value. The letter designations are arbitrary. The numeric portion must be a unique numeric value (unique among all folders and channels) within the XML file; e.g. you cannot have both an "s1" and an "n1" in the same XML file as the "s" and "n" are not stored to the database.
+
+User customizations to a page are represented in the database by "remove ID x" so once a layout is used in production, you must not change the existing ID values or re-use ID values or uPortal will not be able to properly render a user's page that has modified their layout.
+
+After modifying a layout with the Admin UI you should export the fragment-layout and store it in your source code repository so you can recreate the layout in other environments with ant initdb or ant initportal.
+	Y
+fname	Unique name of the portlet created when publishing the portlet.	Y
```
Comment (ChristianMurphy) [inline docs/implement/layout-management/make-a-fragment.md]: Would this be more easy to read as a table?

File: docs/implement/layout-management/make-a-fragment.md [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+Child of a column folder:
+width	percentage
+
+Defines the width of the column in percentage; e.g. 50%. The column widths in a tab/row should add up to 100
+
+Additional References
+
+Overview
+
+Don't forget that in addition to creating the *-fragment-layout.xml file (like https://github.com/Jasig/uPortal/blob/uportal-4.2.0/uportal-war/src/main/data/quickstart_entities/fragment-layout/academics-lo.fragment-layout.xml), you also need to:
```
Comment (ChristianMurphy) [inline docs/implement/layout-management/make-a-fragment.md]: Escape `*` :arrow_right:  `\*`

File: docs/implement/layout-management/make-a-fragment.md [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+
+    Create a corresponding *-fragment-definition.xml file (see https://github.com/Jasig/uPortal/blob/uportal-4.2.0/uportal-war/src/main/data/quickstart_entities/fragment-definition/academics-lo.fragment-definition.xml for an example)
```
Comment (ChristianMurphy) [inline docs/implement/layout-management/make-a-fragment.md]: Escape `*` :arrow_right:  `\*`

**Comment by ChristianMurphy** [comment]
:bell: @apetro

**Comment by ChristianMurphy** [comment]
This seems to be in a pretty good state, merging as is, additional adjustments can come in follow up PRs.
Thanks @apetro! :bow:

---

:notebook: Whenever possible, please create branches on a fork of the `Jasig/uPortal` repo, it keeps the upstream branches pristine, and avoids confusion when I periodically sweep through branches to prune inactive ones.

**Comment by ChristianMurphy** [comment]
> I'm not entirely comfortable with this clearing the WIP status on a contributor's work-in-progress PR and merging it.

I apologize for any discomfort caused.

> The part where the contributor becomes comfortable that, or is convinced that, the contribution is indeed ready to merge, is important.

uPortal has problems with stale and abandoned PRs getting left behind.
After two weeks of no updates on a non-controversial change, and a ping being ignored, I'd consider it abandoned.
Would the preferred resolution be to close and mark as `stale` or merged?

**Comment by apetro** [comment]
To be clear, I've got several mis-steps here to own up to:

+ I probably should have created the branch in a fork rather than in `Jasig/uPortal`, just as you suggest, @ChristianMurphy . There's much to be said for avoiding clutter (as others in unrelated parts of my life often tell me 😄 ) I'm not sure I intentionally navigated that (vs doing whatever GitKrakken defaulted to), but to the extent that I thought about it, I probably had in mind that putting it in upstream better signalled my interest in others collaborating on this documentation. (Wisc doesn't use all the DLM features; someone who does presumably can better wrangle the documentation of; more generally that documentation is seriously stale and needs love.)
+ I can (now) see how 🔔 was or at least could have been meant to imply "This looks good enough, let's merge it and make it more better in subsequent pull requests". This is what emojis are for, shorthand, it is both their strengths and limitations. For whatever it's worth, I didn't get all that nuance from that emoji when it tickled my octobox.io dashboard. What I got is "oh yeah, I should look at that again, I'll leave that item in my octobox.io inbox as a reminder to get to this". And I can see how not responding to it for 2 days felt like "ignoring" it. (I would have sworn I up-thumbed that bell, but apparently not; certainly would have been better to do so as a lightweight way to acknowledge). Objectively it's just a fact that not doing anything in response _is_ ignoring it. But for whatever it's worth, it also _isn't_ ignoring it, in the sense that my timeline for getting back to (some) things is (a lot) longer than 2 days. Too many balls in the air. This documentation is in that nice-to-have category of it came up locally when I needed to explain DLM to someone, which made me notice the documentation I could point them to was less than stellar, which made me open a thread of effort to make that better, and then, well, distractions. :squirrel:
+ Of course, on the substance, I just should have gotten to this more sooner. It's not _that far_ off and dashing some "Stale, needs love" kinds of disclaimers into it would probably mitigate risks in the documentation as-is.

> Would the preferred resolution be to close and marked as stale or merged?

I'd hope a "Hey Andrew this looks pretty good as is, how about you merge this as is and you and others can follow up via subsequent PRs as opportunity allows?" would have gotten the right response from me. Maybe that's wishful thinking, I might not have gotten to engaging with that expeditiously either. 🤷‍♂️ 

Yes, when it comes right down to it, if as a Committer you can't get a contributor to clear a "WIP" designation, I do think the right resolution is to close the PR un-merged with attendant sadness. As a pedantic matter uPortal shouldn't be accepting contributions that aren't unambiguously intended by the contributor to be accepted in the form they come through -- all of the magic of our licensing hinges on contributor intent.

**Comment by ChristianMurphy** [comment]
> I do think the right resolution is to close the PR un-merged with attendant sadness.

:disappointed: Alright

> all of the magic of our licensing hinges on contributor intent.

Opening a PR is clear intent.

**Comment by apetro** [comment]
In general, I agree opening a PR is delightfully clear intent. It's right in the name. "Pull request". As in I request that you pull this. GitHub is a beautiful thing.

As a practical matter, pulling in WIP contributions is never, not once, going to actually be a problem in terms of licensing. In the same way that we don't really need CLAs at all.

And that's probably where I should leave this. There's no substantive problem here: I do appreciate your shaking this contribution loose and I (still!) intend to circle back and improve it when I can get to it. I think I've understood where you're coming from on this (and you're not wrong!), and I think you've heard where I'm coming from on this (thanks for listening!), so I think we got both the substantive and meta goodness there is to be gotten from this PR. On to the next one! :smile:

**Pull request merged**